src/solvers/dgsem_p4est/dg_2d_parabolic.jl

```diff
@@ -0,0 +1,344 @@
+# the RHS etc.
+function create_cache_parabolic(mesh::P4estMesh, equations_hyperbolic::AbstractEquations,
+                                equations_parabolic::AbstractEquationsParabolic,
+                                dg::DG, parabolic_scheme, RealT, uEltype)
+
+  balance!(mesh)
+
+  elements   = init_elements(mesh, equations_hyperbolic, dg.basis, uEltype)
+  interfaces = init_interfaces(mesh, equations_hyperbolic, dg.basis, elements)
+  boundaries = init_boundaries(mesh, equations_hyperbolic, dg.basis, elements)
+ 
+  n_vars = nvariables(equations_hyperbolic)
+  n_elements = nelements(elements)
+  n_nodes = nnodes(dg.basis) # nodes in one direction
+  u_transformed = Array{uEltype}(undef, n_vars, n_nodes, n_nodes, n_elements)
+  gradients = ntuple(_ -> similar(u_transformed), ndims(mesh))
+  flux_viscous = ntuple(_ -> similar(u_transformed), ndims(mesh))
+
+  cache = (; elements, interfaces, boundaries, gradients, flux_viscous, u_transformed)
+
+  return cache
+end
+
+function calc_gradient!(gradients, u_transformed, t,
+                        mesh::P4estMesh{2}, equations_parabolic,
+                        boundary_conditions_parabolic, dg::DG, 
+                        cache, cache_parabolic)
+
+  gradients_x, gradients_y = gradients
+
+  # Reset du
+  @trixi_timeit timer() "reset gradients" begin
+    reset_du!(gradients_x, dg, cache)
+    reset_du!(gradients_y, dg, cache)
+  end
+
+  # Calculate volume integral
+  @trixi_timeit timer() "volume integral" begin
+    (; derivative_dhat) = dg.basis
+    (; contravariant_vectors) = cache.elements
+    
+    @threaded for element in eachelement(dg, cache)
+
+      # Calculate gradients with respect to reference coordinates in one element
+      for j in eachnode(dg), i in eachnode(dg)
+        u_node = get_node_vars(u_transformed, equations_parabolic, dg, i, j, element)
+
+        for ii in eachnode(dg)
+          multiply_add_to_node_vars!(gradients_x, derivative_dhat[ii, i], u_node, equations_parabolic, dg, ii, j, element)
+        end
+
+        for jj in eachnode(dg)
+          multiply_add_to_node_vars!(gradients_y, derivative_dhat[jj, j], u_node, equations_parabolic, dg, i, jj, element)
+        end
+      end
+
+      # now that the reference coordinate gradients are computed, transform them node-by-node to physical gradients
+      # using the contravariant vectors
+      for j in eachnode(dg), i in eachnode(dg)
+        Ja11, Ja12 = get_contravariant_vector(1, contravariant_vectors, i, j, element)
+        Ja21, Ja22 = get_contravariant_vector(2, contravariant_vectors, i, j, element)
+
+        gradients_reference_1 = get_node_vars(gradients_x, equations_parabolic, dg, i, j, element)
+        gradients_reference_2 = get_node_vars(gradients_y, equations_parabolic, dg, i, j, element)
+
+        gradient_x_node = Ja11 * gradients_reference_1 + Ja21 * gradients_reference_2
+        gradient_y_node = Ja12 * gradients_reference_1 + Ja22 * gradients_reference_2
```

Comment:
Is this correct? I'm not sure because the contravariant vector notation is different from the notation I use.

Reply:
@andrewwinters5000?

Reply:
No, this is not correct. To transform a gradient from computational to physical coordinates, you get the i-th component by taking the scalar product of the i-th contravariant vector with the gradient in computational coordinates.
It must be
```julia
gradient_x_node = Ja11 * gradients_reference_1 + Ja12 * gradients_reference_2
gradient_y_node = Ja21 * gradients_reference_1 + Ja22 * gradients_reference_2
```

Derivation (using my notation where I always use superscripts to access components of a vector instead of Kopriva's, which uses a confusing combination of upper and lower indexing):
![grafik](https://github.com/trixi-framework/Trixi.jl/assets/44124897/81730e52-ea64-43d5-9836-572c6e7840bf)

Reply:
This doesn't seem to be correct - the errors are larger and mesh artifacts are more visible with this formula. 
<img width="623" alt="Screenshot 2023-06-01 at 9 43 43 AM" src="https://github.com/trixi-framework/Trixi.jl/assets/1156048/bfe08e54-3819-4a46-b4ea-d058f9a6aa7d">

Compare with results for the current code: 
<img width="606" alt="Screenshot 2023-06-01 at 9 44 07 AM" src="https://github.com/trixi-framework/Trixi.jl/assets/1156048/2bf15aba-58f2-487d-886d-09c5fba9f548">

Reply:
This is weird. I'm pretty sure that's correct, it's the same as for the volume integral. Maybe there's a bug somewhere else?

Reply:
The volume integral is computing a divergence, so I think the matrix of metric terms is transposed when computing a gradient.

Reply:
Why would it be transposed? I've double-checked, and I still don't see why your transposed version would work.

Reply:
Can you explain that, please?
For the divergence, we do that:
![grafik](https://github.com/trixi-framework/Trixi.jl/assets/44124897/0c766129-7438-408e-8984-f6aae8046529)
And then we use a few tricks to get the `n_i` inside the derivative to obtain that:
![grafik](https://github.com/trixi-framework/Trixi.jl/assets/44124897/223b05ef-4ca4-4ea7-a3e2-649984c0f989)
The former is just basic differential geometry, the exact same thing that I did before on the notepad, and the latter doesn't transform anything.

Reply:
Here's the example in 2D using your notation: let F = (f1, f2). Then

div_F = d/dx1 * f1 + d/dx2 * f2
          = dxi1/dx1 * d/dxi1 * f1 + dxi2/dx1 * d/dxi2 * f1 + 
             dxi1/dx2 * d/dxi1 * f2 + dxi2/dx2 * d/dxi2 * f2 

Since d/dxi1 * dxj/dxi1 + d/dxi2 * dxj/dxi2 = 0 for j = 1, 2 due to the geometric conservation law/metric identities, we can pull the geometric terms inside the derivatives so that

div_F = d/dxi1 * (dxi1/dx1 * f1) + d/dxi2 * (dxi2/dx1 * f1) + 
             d/dxi1 * (dxi1/dx2 * f2) + d/dxi2 * (dxi2/dx2 * f2)
          = d/dxi1 * (dxi1/dx1 * f1) + d/dxi1 * (dxi1/dx2 * f2) + 
             d/dxi2 * (dxi2/dx1 * f1) + d/dxi2 * (dxi2/dx2 * f2)

where I just rearranged terms in the last line. 

Note that for the divergence, we're computing the matrix product
[dxi1/dx1 dxi1/dx2]  [f1]
[dxi2/dx1 dxi2/dx2] [f2]
to create a rotated flux in the volume kernel, which we then differentiate with respect to computational coordinates. However, for the gradient, we compute


[dxi1/dx1, dxi2/dx1] [df1/dxi1]
[dxi1/dx2, dxi2/dx2] [df1/dxi2]

This is what I am doing in my implementation as well.

Reply:
I think the main difference between the two is that one is rotating the flux before taking derivatives, while the other has already taken the derivatives and is rotating the result.

Reply:
You're right, I did the correct math, but then confused the definitions of the contravariant vectors.
![grafik](https://github.com/trixi-framework/Trixi.jl/assets/44124897/a88851a4-81ed-4aab-ab90-2bb595ee34a5)
What I did before was correct until I replaced the dxi/dx by entries of n. It should be
![grafik](https://github.com/trixi-framework/Trixi.jl/assets/44124897/b4ec252d-feaa-46aa-9800-c6538be38e34)
Which is then exactly what you did.

Reply:
> I think the main difference between the two is that one is rotating the flux before taking derivatives, while the other has already taken the derivatives and is rotating the result.

The difference between the two is that the divergence F has multiple components, and we take the scalar product with dF/dxi^i (all components, one derivative) and not df/dxi (one component, all derivative), which is kinda the transposed version (if we had multiple components as well, at least).